pkg/webhook/resourcedistribution/validating/utils.go

```diff
@@ -0,0 +1,192 @@
+var (
+	// SupportGVKList is a list that contains all support resource group, version, and kind
+	// Support CustomResourceDefinition
+	/* ADD NEW RESOURCE TYPE HERE*/
+	supportGVKList = []schema.GroupVersionKind{
+		{Group: "", Version: "v1", Kind: "Secret"},
+		{Group: "", Version: "v1", Kind: "ConfigMap"},
+	}
+
+	// Resources will never be distributed to namespaces in this list
+	/* ADD NEW FORBIDDEN NAMESPACE HERE*/
+	forbiddenNamespaces = []string{
+		"kube-system",
+		"kube-public",
+	}
+)
+
+// SupportResourceGVK check whether object is supported by ResourceDistribution
+func SupportResourceGVK(object runtime.Object) bool {
+	if object == nil {
+		return false
+	}
+	objGVK := object.GetObjectKind().GroupVersionKind()
+	for _, gvk := range supportGVKList {
+		if reflect.DeepEqual(gvk, objGVK) {
+			return true
+		}
+	}
+	return false
+}
+
+// IsNotInForbidden check whether the namespace is forbidden
+func IsNotInForbidden(namespace string) bool {
+	for _, forbiddenNamespace := range forbiddenNamespaces {
+		if namespace == forbiddenNamespace {
+			return false
+		}
+	}
+	return true
+}
+
+// ConvertToUnstructured receives runtime.Object, returns *unstructured.Unstructured
+func ConvertToUnstructured(resourceObject runtime.Object) (resource *unstructured.Unstructured) {
+	switch unstructuredResource := resourceObject.(type) {
+	case *unstructured.Unstructured:
+		return unstructuredResource
+	default:
+		return nil
+	}
+}
+
+// DeserializeResource receives yaml of resource, returns UnifiedResource
+func DeserializeResource(resourceRawExtension *runtime.RawExtension, fldPath *field.Path) (resource runtime.Object, allErrs field.ErrorList) {
+	// 1. check whether resource yaml is empty
+	if resourceRawExtension.Raw == nil {
+		return nil, append(allErrs, field.Invalid(fldPath, nil, "empty resource is not allowed"))
+	}
+
+	// 2. deserialize resource
+	resource, _, err := unstructured.UnstructuredJSONScheme.Decode(resourceRawExtension.Raw, nil, nil)
+	if err != nil {
+		allErrs = append(allErrs, field.InternalError(fldPath, fmt.Errorf("failed to deserialize resource, please check your spec.resource, err %v", err)))
+	}
+	return
+}
+
+// GetTargetNamespaces will return **the intersection** of them when multiple options are enabled.
+func GetTargetNamespaces(handlerClient client.Client, targets *appsv1alpha1.ResourceDistributionTargets) (selectedNamespaces []string, allErrs field.ErrorList) {
+	threshold := 0
+	selectedTimes := make(map[string]int)
+
+	// 1. select namespaces via targets.IncludedNamespaces option
+	if len(targets.IncludedNamespaces) != 0 {
+		threshold++
+		for _, namespace := range targets.IncludedNamespaces {
+			selectedTimes[namespace.Name] = 1
+		}
+	}
+
+	// 2. select namespaces via targets.NamespaceLabelSelector option
+	if len(targets.NamespaceLabelSelector.MatchLabels) != 0 || len(targets.NamespaceLabelSelector.MatchExpressions) != 0 {
+		selectors, err := metav1.LabelSelectorAsSelector(&targets.NamespaceLabelSelector)
```

Comment:
consider using GetFastLabelSelector to speedup selection in some cases

Reply:
ok